plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

```diff
@@ -1490,30 +1494,50 @@ private static void cleanUpGlobalClassValue(@NonNull ClassLoader loader) throws
-        while (it.hasNext()) {
-            Class<?> klazz = it.next();
-            ClassLoader encounteredLoader = klazz.getClassLoader();
-            if (encounteredLoader != loader) {
-                it.remove();
-                if (LOGGER.isLoggable(Level.FINEST)) {
-                  LOGGER.finest(() -> "ignoring " + klazz + " with loader " + encounteredLoader);
-                }
-            }
-        }
+        removeClassesFromOtherClassLoaders(it, loader);
         LOGGER.fine(() -> "cleaning up " + toRemove + " associated with " + loader);
         for (Class<?> klazz : toRemove) {
+            loadedClasses.add(klazz);
             removeM.invoke(map, klazz);
         }
     }
 
-    private static void cleanUpClassHelperCache(@NonNull Class<?> clazz) throws Exception {
+    private static void cleanUpClassHelperCache(@NonNull ClassLoader loader, Set<Class<?>> loadedClasses) throws Exception {
         Field classCacheF = Class.forName("org.codehaus.groovy.ast.ClassHelper$ClassHelperCache").getDeclaredField("classCache");
         classCacheF.setAccessible(true);
         Object classCache = classCacheF.get(null);
-        if (LOGGER.isLoggable(Level.FINER)) {
-            LOGGER.log(Level.FINER, "cleaning up {0} from ClassHelperCache? {1}", new Object[] {clazz.getName(), classCache.getClass().getMethod("get", Object.class).invoke(classCache, clazz) != null});
+        Class<?> classCacheC = classCache.getClass();
+        Collection entries = (Collection) classCacheC.getMethod("values").invoke(classCache);
+        Class<?> managedRefC = Class.forName("org.codehaus.groovy.util.ManagedReference");
+        var getRefM = managedRefC.getMethod("get");
+        List<Class<?>> toRemove = new ArrayList<>(); // not sure if it is safe against ConcurrentModificationException or not
+        for (Object entry : entries) {
+            Class<?> clazz = (Class<?>) getRefM.invoke(entry);
+            if (clazz != null) {
+                toRemove.add(clazz);
+            }
+        }
+        Iterator<Class<?>> it = toRemove.iterator();
+        removeClassesFromOtherClassLoaders(it, loader);
+        LOGGER.fine(() -> "cleaning up " + toRemove + " associated with " + loader);
+        Method removeM = classCache.getClass().getMethod("remove", Object.class);
+        for (Class<?> klazz : toRemove) {
+            loadedClasses.add(klazz);
+            removeM.invoke(classCache, klazz);
+        }
+    }
+
+    private static void removeClassesFromOtherClassLoaders(Iterator<Class<?>> it, ClassLoader loader) {
+        while (it.hasNext()) {
+            Class<?> klazz = it.next();
+            ClassLoader encounteredLoader = klazz.getClassLoader();
+            if (encounteredLoader != loader) {
+                it.remove();
+                if (LOGGER.isLoggable(Level.FINEST)) {
+                  LOGGER.finest(() -> "ignoring " + klazz + " with loader " + encounteredLoader);
+                }
```

Comment:
```suggestion
                LOGGER.finest(() -> "ignoring " + klazz + " with loader " + encounteredLoader);
```

I guess (already using the `Supplier` overload)